crux-lucene/src/crux/lucene.clj

```diff
@@ -174,79 +172,142 @@
+  (let [directory (if db-dir
+                    (FSDirectory/open db-dir)
+                    (ByteBuffersDirectory.))
+        index-writer (->index-writer {:directory directory, :analyzer analyzer,
+                                      :index-deletion-policy (SnapshotDeletionPolicy. (KeepOnlyLastCommitDeletionPolicy.))})
         searcher-manager (SearcherManager. index-writer false false nil)
-        lucene-store (LuceneNode. directory analyzer index-writer searcher-manager indexer)]
+        cp-job (when checkpointer
+                 (cp/start checkpointer (checkpoint-src index-writer) {::cp/cp-format "lucene-8"}))
+        lucene-store (LuceneNode. directory analyzer
+                                  index-writer searcher-manager
+                                  indexer
+                                  cp-job
+                                  (doto (.newThread (cio/thread-factory "crux-lucene-fsync")
+                                                    #(fsync-loop index-writer fsync-frequency))
+                                    (.start)))]
+
     ;; Ensure lucene index exists for immediate queries:
     (.commit index-writer)
-    (validate-lucene-store-up-to-date index-store lucene-store)
     (q/assoc-pred-ctx! query-engine ::lucene-store lucene-store)
-    (bus/listen bus {:crux/event-types #{:crux.tx/committing-tx :crux.tx/aborting-tx}
-                     :crux.bus/executor (reify java.util.concurrent.Executor
-                                          (execute [_ f]
-                                            (.run f)))}
-                (fn [ev]
-                  (when (= :crux.tx/committing-tx (:crux/event-type ev))
-                    (index! indexer index-writer (db/fetch-docs document-store (:doc-ids ev)))
-                    (when-let [evicting-eids (not-empty (:evicting-eids ev))]
-                      (evict! indexer index-writer evicting-eids)))
-                  (index-tx! index-writer (:submitted-tx ev))
-                  (.commit index-writer)
-                  (.maybeRefreshBlocking searcher-manager)))
+
+    (tx/register-index! secondary-indices
+                        (latest-completed-tx-id index-writer)
+                        (fn [{:keys [::tx/tx-id ::txe/tx-events committing?]}]
+                          (when committing?
+                            (let [{:keys [docs evicted-eids]} (transform-tx-events document-store tx-events)]
+                              (when-let [evicting-eids (not-empty evicted-eids)]
+                                (evict! indexer index-writer evicting-eids))
+                              (index! indexer index-writer docs)))
+
+                          (.setLiveCommitData index-writer {"crux.tx/tx-id" (str tx-id)
+                                                            "crux.lucene/index-version" (str index-version)})
+                          (.maybeRefreshBlocking searcher-manager)))
```

Comment:
In the other PR I debated whether to include this as part of the `l/await-tx` call https://github.com/juxt/crux/pull/1535#discussion_r653487129

...but now we're back to ~sync indexing I think it's fine/necessary to stay living here :+1: